File: src/main/java/Manchkin/game/Box.java
package Manchkin.game;


import java.util.*;

public class Box {
    // В коробке стоит стопка
    public ArrayList<Card> Chunk = new ArrayList<>();

    // Положить в коробку карту
    public void add(Card card, Card.Place place, int x, int y, int player){
        card.move(place, x, y, player);
        Chunk.add(card);
    }

    // Взять из коробки карту
    public Card remove(int number){
        Card card;
        card = findCard(number);
        Chunk.remove(card);
        Chunk.trimToSize();
        return card;
    }

    public Card removeTop(){
        Card card;
        card = Chunk.get(Chunk.size() - 1);
        Chunk.remove(Chunk.size() - 1);
        Chunk.trimToSize();
        return card;
    }

    // Найти в коробке карту
    protected Card findCard(int number){
        for(Card card : Chunk){
            if(card.number == number)
                return card;
        }
        System.out.println("Создана несуществующая карта в Box");
        return new Card(-1, "", "", 0);
    }
}

File: src/main/webapp/JavaScript/game/Box.js

function Box(type, $Garbage){
    this.type = type;
    this.$Garbage = $Garbage;
    this.chunk = new Array();
}

Box.prototype.createCard = function(number, player, place, type, sprite, x, y, rotate){
    if(!this.chunk.hasOwnProperty(number)) {
        var $owner = this.find$Place(place, player);
        this.chunk[number] = new Card($owner, number, player, place, type, sprite, x, y, $owner.dx, $owner.dy, rotate);
    }
};

Box.prototype.moveCard = function(number, place, player, x, y, rotate){
    if(!this.chunk.hasOwnProperty(number)) alert("Карта не создана");
    else{
        var $owner = this.find$Place(place, player);
        this.chunk[number].move($owner, player, place, x, y, $owner.dx, $owner.dy, rotate);
    }
};

Box.prototype.removeCard = function(number){
    var typeOfCard = this.type;
    $('.' + typeOfCard + '[number = ' + number + ']').remove();
    if(this.chunk.hasOwnProperty(number)){
        this.chunk.splice(number);
    }
};

Box.prototype.find$Place = function(place, player){
    var $owner;
    switch (place){
        case 'Table':
            $owner = Table.$table.$place;
            break;
        case 'PlayerHand':
            $owner = players[player].$hand.$place;
            break;
        case 'PlayerTable':
            $owner = players[player].$table.$place;
            break;
        case 'Garbage':
            $owner = this.$Garbage.$place;
            break;
        default : alert("Попытка переместится в неизвестное место " + place);
    }
    return $owner;
}

Box.prototype.sizing = function(){
    for(var card in this.chunk){
        if(!this.chunk.hasOwnProperty(card))    continue;
        if(this.chunk[card].place != 'PlayerHand' && this.chunk[card].place != 'PlayerTable')   continue;
        var it = this.chunk[card];
        var $owner = this.find$Place(it.place, it.player);
        it.moveLight($owner.dx, $owner.dy);
    }
};

function $Place($place, place, player){
    this.$place = $place;

    this.$place.on('mousedown', 'div', {place: place, player: player}, click);
    this.$place.on('mousedown', {place: place, player: player}, Move.down);
    this.$place.on('mouseleave', 'div', small);
    this.$place.on('contextmenu', 'div', function(){
        return false;
    });

}

$Place.prototype.sizing = function(){
    this.$place.dx = (this.$place.width() - 60)/100;
    this.$place.dy = (this.$place.height() - 94)/100;
};